schedule/schedule.go

```diff
@@ -0,0 +1,189 @@
+package schedule
+
+import (
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/ghodss/yaml"
+)
+
+const (
+	DateFormat = "Mon 02 Jan 2006"
+)
+
+// Schedule represents a series of Shifts, in temporal order.
+type Schedule struct {
+	// Shifts is the list of shifts in order. The last Shift, and only the last Shift, should have a StopTime value.
+	Shifts []*Shift `json:"shifts"`
+}
+
+func (sch *Schedule) LastShift() *Shift {
+	if sch.Shifts == nil || len(sch.Shifts) == 0 {
+		return nil
+	}
+	return sch.Shifts[len(sch.Shifts)-1]
+}
+
+func (sch *Schedule) Validate() error {
+	if sch == nil {
+		return fmt.Errorf("schedule cannot be nil")
+	}
+
+	if sch.Shifts == nil || len(sch.Shifts) == 0 {
+		fmt.Println("No input shifts found. Starting from scratch.")
+		return nil
+	}
+
+	var previousStartDate time.Time
+	for i, shift := range sch.Shifts {
+		if err := shift.Validate(); err != nil {
+			return fmt.Errorf("schedule is invalid because shift %v is invalid: %v", i, err)
+		}
+
+		if !previousStartDate.IsZero() && previousStartDate.After(shift.StartDate) {
+			return fmt.Errorf("shifts out of order: shift time %v found after %v", previousStartDate, shift.StartDate)
+		}
+		previousStartDate = shift.StartDate
+
+		if shift == sch.LastShift() {
+			if shift.StopDate.IsZero() {
+				return fmt.Errorf("missing stop date on last shift")
+			}
+		} else {
+			// All other shifts should not have a stop date.
+			if !shift.StopDate.IsZero() {
+				return fmt.Errorf("stop date is only valid on the last shift. Found stop date %v on shift at index %v", shift.StopDate, i)
+			}
+		}
+	}
+
+	return nil // It's all good.
+}
+
+func (sch *Schedule) String() string {
+	if sch == nil {
+		return ""
+	}
+	b, err := yaml.Marshal(sch)
+	if err != nil {
+		return err.Error()
+	}
+	return string(b)
+}
+
+// A shift represents a span of time someone is on duty.
+type Shift struct {
+	// The start date, inclusive, of when the User goes on duty. Only year, month, and day field are relevant.
+	StartDate time.Time `json:"startDate,omitempty"`
+	User      string    `json:"user,omitempty"`
+
+	// UserOverride should be used when Users want to swap or alter shift assignments without disturbing the normal
+	// rotation cycle.
+	UserOverride string `json:"userOverride,omitempty"`
+
+	// StopDate is inclusive, and must only be used on the last Shift of a Schedule. For all other Shifts, the stop date
+	// is implied by the next shift's StartDate, and this value should remain the zero `time.Time` value.
+	StopDate time.Time `json:"stopDate,omitempty"`
```

Comment:
Would it better reflect the data model if the `StopDate` field were on the schedule itself?  Then a whole class of invalid states would no longer be representable.

Also, not to completely revisit the data model...but what if a `Schedule` had a `StartDate` and an ordered array of `Shift`s each of which had:
* `User`
* `UserOverride`
* `Length` as a `Duration`

I think then it would be impossible to represent invalid states in the data model, and most of the validation could go away.

Reply:
Remember the intent of this is to be human readable and editable. A single start date and a series of numbers to add to that start date is difficult to see "when am I on rotation next?"

Reply:
Wouldn't the "when am I on rotation next" be answered by looking at the calendar this is pushing to?  Or do we want to also have the data model be such that people can use that directly to see when they're on rotation?

Reply:
That's one way, but if you were to want to swap a rotation you would go to the schedule file and override your selected shift